src/OpenTelemetry/Metrics/Exemplar/SimpleFixedSizeExemplarReservoir.cs

```diff
@@ -1,140 +1,56 @@
-    private readonly Random random;
-    private readonly Exemplar[] runningExemplars;
-    private readonly Exemplar[] tempExemplars;
+    private readonly Random random = new();
 
-    private long measurementsSeen;
+    private int measurementsSeen;
 
     public SimpleFixedSizeExemplarReservoir(int poolSize)
+        : base(poolSize)
     {
-        this.poolSize = poolSize;
-        this.runningExemplars = new Exemplar[poolSize];
-        this.tempExemplars = new Exemplar[poolSize];
-        this.measurementsSeen = 0;
-        this.random = new Random();
     }
 
-    public override void Offer(long value, ReadOnlySpan<KeyValuePair<string, object?>> tags, int index = default)
+    public override void Offer(in ExemplarMeasurement<long> measurement)
     {
-        this.Offer(value, tags);
+        this.Offer(in measurement);
     }
 
-    public override void Offer(double value, ReadOnlySpan<KeyValuePair<string, object?>> tags, int index = default)
+    public override void Offer(in ExemplarMeasurement<double> measurement)
     {
-        this.Offer(value, tags);
+        this.Offer(in measurement);
     }
 
-    public override Exemplar[] Collect(ReadOnlyTagCollection actualTags, bool reset)
+    protected override void OnCollected()
     {
-        for (int i = 0; i < this.runningExemplars.Length; i++)
-        {
-            this.tempExemplars[i] = this.runningExemplars[i];
-            if (this.runningExemplars[i].FilteredTags != null)
-            {
-                // TODO: Better data structure to avoid this Linq.
-                // This is doing filtered = alltags - storedtags.
-                // TODO: At this stage, this logic is done inside Reservoir.
-                // Kinda hard for end users who write own reservoirs.
-                // Evaluate if this logic can be moved elsewhere.
-                // TODO: The cost is paid irrespective of whether the
-                // Exporter supports Exemplar or not. One idea is to
-                // defer this until first exporter attempts read.
-                this.tempExemplars[i].FilteredTags = this.runningExemplars[i].FilteredTags!.Except(actualTags.KeyAndValues.ToList()).ToList();
-            }
-
-            if (reset)
-            {
-                this.runningExemplars[i].Timestamp = default;
-            }
-        }
-
         // Reset internal state irrespective of temporality.
         // This ensures incoming measurements have fair chance
         // of making it to the reservoir.
-        this.measurementsSeen = 0;
-
-        return this.tempExemplars;
+        Interlocked.Exchange(ref this.measurementsSeen, 0);
     }
 
-    private void Offer(double value, ReadOnlySpan<KeyValuePair<string, object?>> tags)
+    private void Offer<T>(in ExemplarMeasurement<T> measurement)
+        where T : struct
     {
-        if (this.measurementsSeen < this.poolSize)
+        var measurementNumber = Interlocked.Increment(ref this.measurementsSeen) - 1;
```

Comment:
Could we use regular statements instead of interlocked statements for this PR? The exemplar offer and collect are anyway synchronized under the lock used by MetricPoint update and snapshot. I'm asking this mainly because it would be simpler to review all the thread synchronization related changes in one single PR instead of having them scattered across PRs.

Reply:
How sure are you about this? `AcquireLock` and `ReleaseLock` synchronize the `int isCriticalSectionOccupied` on `MetricPointOptionalComponents`. Threads writing measurements are updating `SimpleFixedSizeExemplarReservoir.measurementsSeen` and the collection thread is updating `SimpleFixedSizeExemplarReservoir.measurementsSeen`. It needs its own synchronization, no?

Reply:
> Threads writing measurements are updating SimpleFixedSizeExemplarReservoir.measurementsSeen and the collection thread is updating SimpleFixedSizeExemplarReservoir.measurementsSeen. It needs its own synchronization, no?

These operations are still being done under the same lock on `MetricPointOptionalComponents.isCriticalSectionOccupied`, right?

- Update thread -> acquire lock -> make the update -> call `ExemplarReservoir.Offer` -> release lock
- Snapshot thread -> acquire lock -> take the snapshot -> call `ExemplarReservoir.Collect` -> release lock

Reply:
I don't _think_ so. But I could be wrong 😄 If we were using `lock` aka `Monitor` than yes. I would agree. But we're not using a true "lock" we're just manually synchronizing some int to simulate a lock. Because of that I think we also need to synchronize `measurementsSeen`.

Reply:
> If we were using lock aka Monitor than yes. I would agree. But we're not using a true "lock" we're just manually synchronizing some int to simulate a lock.

For our requirements, what we need is way to ensure that at any given point, there cannot be more than one thread that executes the critical section. The synchronization mechanism that we use ensures that. Both the traditional `Monitor/lock` and our implementation of the spin lock fulfil that requirement equally well. There's no question of stronger or weaker lock here.

Reply:
OK updated to remove the new interlocked bits.